crates/ty_server/src/server/api/diagnostics.rs

```diff
@@ -64,38 +63,37 @@ pub(super) fn clear_diagnostics(key: &DocumentKey, client: &Client) {
 /// This function is a no-op if the client supports pull diagnostics.
 ///
 /// [publish diagnostics notification]: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_publishDiagnostics
-pub(super) fn publish_diagnostics(
-    session: &Session,
-    key: &DocumentKey,
-    client: &Client,
-) -> Result<()> {
+pub(super) fn publish_diagnostics(session: &Session, key: &DocumentKey, client: &Client) {
     if session.client_capabilities().pull_diagnostics {
-        return Ok(());
+        return;
     }
 
     let Some(url) = key.to_url() else {
-        return Ok(());
+        return;
     };
 
-    let path = key.path();
+    let snapshot = session.take_document_snapshot(url.clone());
 
-    let snapshot = session
-        .take_document_snapshot(url.clone())
-        .ok_or_else(|| anyhow::anyhow!("Unable to take snapshot for document with URL {url}"))
-        .with_failure_code(lsp_server::ErrorCode::InternalError)?;
+    let document = match snapshot.document() {
+        Ok(document) => document,
+        Err(err) => {
+            tracing::debug!("Failed to resolve document for URL `{}`: {}", url, err);
+            return;
+        }
+    };
 
-    let db = session.project_db_or_default(path);
+    let db = session.project_db_or_default(key.path());
 
     let Some(diagnostics) = compute_diagnostics(db, &snapshot) else {
```

Comment:
I suggest passing the document in addition to the snapshot so that `compute_diagnostic` doesn't have to repeat the `document` dance

Reply:
This function is used in https://github.com/astral-sh/ruff/blob/08c2c695a2006cf2d141478a76822f0ce755fc0d/crates/ty_server/src/server/api/requests/diagnostic.rs#L42-L42 as well which would mean that the `snapshot > document > file` would need to be done somewhere - either in the document diagnostic handler or inside compute diagnostic.